rfcs/20240718-synthetic-benchmark/README.md

```diff
@@ -0,0 +1,279 @@
+* Developer supplied filters: Developer supplied filters are expected to fit
+  into two categories, value sets or dimension restrictions. Value sets will
+  just be lists of allowed values such as the data type configuration of
+  `{s8:s8:s32,u8:u8:u32}`. Dimension restriction, on the other handle, will be
+  composed of range and modulo restrictions, for example, `{m in [1,32]:
+  m%4==0}`. In addition, as most of the development complexity arises around
+  small dimensions sizes (and large problems take longer to benchmark), indices
+  are expected to be sampled from a power law distribution.
+
+* Automated Diffing: As one of the relevant metrics is a comparison between
+  oneDNN builds, this tool needs a way to execute multiple oneDNN libraries. As
+  a consequence, we can add a testing API for comparing implementations. When
+  implementations match between oneDNN builds, we can often skip benchmarking.
+  As oneDNN primitive creation and dispatching can be quite complicated, this is
+  beneficial for developers as they cannot always predict the effect of an
+  implementation change. In addition, if functionality validation is supported,
+  this provides a much faster way to initially validate correctness by skipping
+  unchanged implementations.
+
+## Proposal
+To address the above issue, this RFC propose creating tools to support a data
+pipeline. For the purposes of this RFC, the pipeline is being split into two
+stages, a data producer and some data consumers. Often data pipelines include a
+data storage step via a database. While this would be useful, it is being
+omitted from the current RFC to reduce the complexity.
+
+### Data collection
+Given the above discussion, this data collection utility is expected to have the
+syntax:
+
+```Bash
+collectdnn --<primitive> --validate --report=<metric> --samples=<num> --<filter_cfg1>=<param1> --<filter_cfg2>=<param2> ...
+```
+
+As a psuedo-code summary for how this will be implemented,
+
+
+```C++
+Generator G(primtitive_kind, filter_cfg, enable_auto_diff);
+Results R;
+std::pair<Problem, Primitive> P;
+parallel_for(Problem p : G) {
+    P.emplace_back({p, create_primitive(p)});
+}
+
+for(auto p : P) {
+    if (benchmark) r.emplace_back(benchmark(p, metric, ctx));
+    if(validate) p.validate();
+}
+
+```
+
+As this tool is intended to produce data from multiple oneDNN builds, this tool
+should rely on benchdnn for data generation. To properly support data
+collection, benchdnn will need two new features. The first feature is to enable
+logging and skipping problems based off of a primitive ID. These IDs can then be
+used for the automated diffing feature discussed above.
+
+There are two main options for implementing primtive IDs
+ * Option 1: Serialize relevant primitive data
+ * Option 2: Hash relevant primitive data to generate a UUID
+
+Given that we are looking at generating millions of primitives, and serialized
+primitive data is likely to be in the 10's of kilobytes in size, Option 1
```

Comment:
There is an API to query a cache blob ID from primitive descriptors to distinguish different primitives. Would it work here and have checked how big it is?

Reply:
I had not considered that API. The biggest issue I see is that primitive descriptors do not contain enough information. For example, consider the recent GEMM fix in commit 25f6896. This only effects the GPU kernel binary generated by the jit:gemm implementation, and so a kernel changed by this commit cannot be caught within the scope of the primitive descriptor.

On the other hand, the previous change in commit e8e05b4 would effect the primitive descriptor, so in theory could be caught by a cache blob ID (with appropriate modifications for this task). I had considered having a mode the filters on primitive descriptor changes, which boils down to a performance/accuracy tradeoff that can use based the work performed. I decided to not to include it until we know generating primitives is too slow in practice. While this definitely an potential issue as we can only generate around ~100 OpenCL kernels per second for the GPU, this shouldn't be an issue with the move to reusable implementations unless searching lots of post-ops combinations becomes important.